### PR TITLE
#22756 fix setting categories via rest api

### LINKED ISSUE #22756: Removing Categories through REST API
**Describe the bug**
Whenever REST API call made to clear/remove categories it returns 500

**To Reproduce**
Example. 

News has category **type** so Today news has two categories sport, entertainment so when trying to remove all categories with **type:{}** It returns 500
/api/v1/workflow/actions/fire
{ 
   "actionName": "Save",
   "comments": "saving content",
   "contentlet": {
      "contentType":"News",
        "identifier": "f3fd1b2042418d84592334b3b9a27627",
        "inode": "22540f01-7322-4c79-bb1a-e57810a85f0b",
        "type":{},
      "title":"blkjjhj Second Content", 
      "languageId": "1"
  }
}.

Logs:

`ERROR workflow.WorkflowResource: Exception on firing, workflow action: ceca71a0-deee-4999-bd47-b01baa1bcfc8, inode: null
java.lang.NullPointerException: null
at com.dotmarketing.portlets.categories.business.CategoryAPIImpl.findByVariable_aroundBody72(CategoryAPIImpl.java:787) ~[dotcms_22.03.2_999999.jar:?]
at com.dotmarketing.portlets.categories.business.CategoryAPIImpl$AjcClosure73.run(CategoryAPIImpl.java:1) ~[dotcms_22.03.2_999999.jar:?]
at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]`

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
### Proposed Changes
* Expected values as categories can be Array of String or comma separated string
* On invalid params expect BadRequest

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java [hunk 1]
```diff
@@ -332,46 +339,99 @@ public List<Category> getCategories (final Contentlet contentlet, final User use
...
+                continue;
+            }
+            if (object == null && map.containsKey(field.variable()) ){
+                //something was set to null a cat on purpose
+                //The requirement says that null must not even be considered
+                //We should only remove a category if an empty array is passed
+                map.remove(field.variable());
+                continue;
+            }
+            if(object == null){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java:377]: this if does not make sense? I think it is safe to remove

File: dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java [hunk 2]
```diff
@@ -332,46 +339,99 @@ public List<Category> getCategories (final Contentlet contentlet, final User use
...
+    /**
+     * given a List and a string representation of a category
+     * This will try to find different criterion to get the respective category
+     * @param categories
+     * @param stringValue
+     * @param user
+     * @param respectFrontendRoles
+     */
+    private void buildUpCategories(final List<Category> categories, final String stringValue,
+            final User user, boolean respectFrontendRoles) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java:398]: set to final

File: dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java [hunk 3]
```diff
@@ -332,46 +339,99 @@ public List<Category> getCategories (final Contentlet contentlet, final User use
...
+        if (UtilMethods.isSet(stringValue)) {
+            final CategoryAPI categoryAPI = APILocator.getCategoryAPI();
+            for (final String categoryValue : stringValue.split("\\s*,\\s*")) {
+
+                    // try it as catId
+                    Category category = Try.of(()->categoryAPI.find(categoryValue, user, respectFrontendRoles)).getOrNull();
+                    if (category != null && InodeUtils.isSet(category.getCategoryId())) {
+                        categories.add(category);
+                        Logger.info(MapToContentletPopulator.class, String.format(" value [%s] resolved as a valid Category-Inode.", categoryValue));
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java:407]: should be info? isn't too noise

**Review by jdotcms — APPROVED** [review]
Great job buddy

File: dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java (hunk 3 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java:407]: honestly, I don't think there will be a ton of requests like this to generate such noise. 
In fact, the line appears just once per request. And it's kinda helpful

Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java:407]: I'm not sure @fabrizzio-dotCMS if it's true using a custom dataset. This log is printed for each category and internally this method is called by ContentResource and WorkflowResource

Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java]: I lowered  the log leve to debug

**Review by nollymar — APPROVED** [review]

**Pull request merged**